Support PointCloud dataset by DatumaroBinary format

vinnamkim opened the pull request:
### Summary
 - Ticket no. 104650
 - Finish to support `PointCloud` dataset by `DatumaroBinary` format for compatibility with `Datumaro` format.

### How to test
<!-- Describe the testing procedure for reviewers, if changes are
not fully covered by unit tests or manual testing can be complicated. -->

### Checklist
<!-- Put an 'x' in all the boxes that apply -->
- [x] I submit my changes into the `develop` branch
- [ ] I have added description of my changes into [CHANGELOG](https://github.com/openvinotoolkit/datumaro/blob/develop/CHANGELOG.md)
- [ ] I have updated the [documentation](
  https://github.com/openvinotoolkit/datumaro/tree/develop/docs) accordingly
- [x] I have added tests to cover my changes
- [x] I have [linked related issues](
  https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword)

### License

- [x] I submit _my code changes_ under the same [MIT License](
  https://github.com/openvinotoolkit/datumaro/blob/develop/LICENSE) that covers the project.
  Feel free to contact the maintainers if that's a concern.
- [x] I have updated the license header for each file (see an example below)

```python
# Copyright (C) 2023 Intel Corporation
#
# SPDX-License-Identifier: MIT
```

vinnamkim committed:
Initial copy from datumaro to datumaro_binary

Signed-off-by: Kim, Vinnam <vinnam.kim@intel.com>

vinnamkim committed:
Some refactoring for Datumaro format

Signed-off-by: Kim, Vinnam <vinnam.kim@intel.com>

vinnamkim committed:
Refactor Datumaro format tests

Signed-off-by: Kim, Vinnam <vinnam.kim@intel.com>

vinnamkim committed:
Merge remote-tracking branch 'upstream/develop' into refactor/datumaro-format

vinnamkim committed:
Update CHANGELOG.md

Signed-off-by: Kim, Vinnam <vinnam.kim@intel.com>

vinnamkim committed:
Fix CHANGELOG.md

Signed-off-by: Kim, Vinnam <vinnam.kim@intel.com>

vinnamkim committed:
Add missing file

Signed-off-by: Kim, Vinnam <vinnam.kim@intel.com>

vinnamkim committed:
Merge remote-tracking branch 'upstream/develop' into refactor/datumaro-format

vinnamkim committed:
Add CommonSemanticSegmentationWithSubsetDirsImporter

Signed-off-by: Kim, Vinnam <vinnam.kim@intel.com>

vinnamkim committed:
Add CommonSemanticSegmentationWithSubsetDirsImporter

 - Refactor CommonSemanticSegmentation unit tests as well

Signed-off-by: Kim, Vinnam <vinnam.kim@intel.com>

vinnamkim committed:
Add Datumaro binary format (only header part)

Signed-off-by: Kim, Vinnam <vinnam.kim@intel.com>

vinnamkim committed:
Add Mappers

 - Support DatasetItem by Datumaro binary format but annotations are not yet.

Signed-off-by: Kim, Vinnam <vinnam.kim@intel.com>

vinnamkim committed:
Add annotation mappers

Signed-off-by: Kim, Vinnam <vinnam.kim@intel.com>

vinnamkim committed:
Fix to use little endian for Mappers

Signed-off-by: Kim, Vinnam <vinnam.kim@intel.com>

vinnamkim committed:
Refactor unit tests

Signed-off-by: Kim, Vinnam <vinnam.kim@intel.com>

vinnamkim committed:
Refactor media path context

Signed-off-by: Kim, Vinnam <vinnam.kim@intel.com>

vinnamkim committed:
Support PointCloud by DatumaroBinary format

Signed-off-by: Kim, Vinnam <vinnam.kim@intel.com>

vinnamkim committed:
Activate the remainder tests which was disabled before

Signed-off-by: Kim, Vinnam <vinnam.kim@intel.com>

vinnamkim committed:
Merge remote-tracking branch 'upstream/develop' into feature/add-datumaro-binary-format-dataset-item

Signed-off-by: Kim, Vinnam <vinnam.kim@intel.com>

vinnamkim committed:
Fix wrong merge

Signed-off-by: Kim, Vinnam <vinnam.kim@intel.com>

vinnamkim committed:
Fix wrong merge again

Signed-off-by: Kim, Vinnam <vinnam.kim@intel.com>

vinnamkim committed:
Quick fix for CI failure because of dill

Signed-off-by: Kim, Vinnam <vinnam.kim@intel.com>

vinnamkim committed:
Merge remote-tracking branch 'upstream/develop' into feature/support-point-cloud-by-datumaro-binary-format

Signed-off-by: Kim, Vinnam <vinnam.kim@intel.com>

vinnamkim committed:
Remove unused line

Signed-off-by: Kim, Vinnam <vinnam.kim@intel.com>

wonjuleee commented on datumaro/plugins/data_formats/datumaro/exporter.py:
```diff
@@ -64,68 +65,89 @@ def items(self):
             image = item.media_as(Image)
             path = image.path
+
             if self._context._save_media:
```
previously, when self._context is given by IDataset, was this correctly operated? The reason why I am asking is, IDataset doesn't have _save_media.

vinnamkim replied:
Previously I wrongly typed it as `IDataset`. `Exporter` is correct one. Python `typing` does not affect to the code behavior. So, the code worked correctly but typing was incorrect.

wonjuleee replied:
Thank you for the clarification.

wonjuleee commented on datumaro/plugins/data_formats/datumaro/exporter.py:
```diff
@@ -64,68 +65,89 @@ def items(self):
+                # Temporarily update image path and save it.
+                image._path = osp.join(
+                    self._context._images_dir, item.subset, self._context._make_image_filename(item)
                 )
```
is `_path` private variable of `MediaElement`? how about using `setter` and `getter` for path?

wonjuleee replied:
It seems like an immutable variable after initialization.

vinnamkim replied:
Yes `path` was designed to be immutable but in this case, we have to make it mutable.
I think that the reason that we use `setter` in object oriented programming is when some member variables are dependent on the other variables. For example,

```python
class Whether:
def __init__(self, celsius: int):
     self._celsius = celsius
     self._text = "hot" if celsius > 30 else "warm"

@setter
def celsius(self, celsius: int):
     self._celsius = celsius
     self._text = "hot" if celsius > 30 else "warm"
```

You can see that `text` is dependent on `celsius` in this example. So, I decided to modify `_path` rather than creating the setter of it because it just increases the code lines like
```python
@setter
def path(self, path: str):
    self._path = path
```

wonjuleee replied:
okay, thanks.

wonjuleee commented on datumaro/plugins/data_formats/datumaro/exporter.py:
```diff
@@ -64,68 +65,89 @@ def items(self):
+
+                # Temporarily update pcd related images paths and save them.
+                for i, img in enumerate(sorted(pcd.extra_images, key=lambda v: v.path)):
+                    img.__path = img.path
```
I am sorry that I was failed to find `__path`. Could you let me know?

vinnamkim replied:
This variable is used to save the original path temporariliy.

wonjuleee commented on datumaro/plugins/data_formats/datumaro/exporter.py:
```diff
@@ -64,68 +65,89 @@ def items(self):
-                                item.id,
-                                ri_desc["path"],
-                            )
-                        )
```
We may not need to save the image if the image has data. Is this correct reason you removed?

vinnamkim replied:
No all the things to save the media (images and point cloud) go into
```python
@contextmanager
    def save_media(self, item: DatasetItem):
```

wonjuleee commented on datumaro/plugins/data_formats/datumaro/exporter.py:
```diff
@@ -64,68 +65,89 @@ def items(self):
-
-        if isinstance(item.media, Image):
+    @contextmanager
+    def save_media(self, item: DatasetItem):
```
We usually use `save_media` for boolean variable to save or not. How about changing the name for avoiding confusions?

vinnamkim replied:
Is it ok `context_save_media`?

wonjuleee replied:
it looks good to me.

vinnamkim replied:
Renamed it.

wonjuleee commented on datumaro/plugins/data_formats/datumaro_binary/mapper/media.py:
```diff
@@ -0,0 +1,113 @@
+
+class MediaMapper(Mapper):
+    @classmethod
+    def forward(cls, obj: Optional[MediaElement]) -> bytes:
```
I am a bit confused that we are going to forward only the media.path through this. Correct?

wonjuleee replied:
If so, how about forwarding media.path instead of separating media types?

vinnamkim replied:
This acts a kind of factory for all kind of media types.
So, it has to write the media `type` first then goes on to write its data structure (including `path`, and etc....) subsequently.

wonjuleee replied:
I meant that we may be possible to locate forwarding `path` at here for all media types first, then the others (like size) can be forwarded according to each media types. What do you think?

vinnamkim replied:
I added `backward_dict()` to `MediaElementMapper` and make `ImageMapper` and `PointCloudMapper` to inherit `MediaElementMapper` to use the common function `backward_dict()`. It will be a way to handle common member variable `path` across all media classes.

wonjuleee commented on datumaro/plugins/data_formats/datumaro_binary/mapper/media.py:
```diff
@@ -0,0 +1,113 @@
+        media_type, height, width = struct.unpack_from("<Iii", _bytes, offset)
+        assert media_type == cls.MEDIA_TYPE
+        size = (height, width)
+        offset += 12
```
Could you clarify why this requires 12 bytes?

vinnamkim replied:
In L80, you can see we pack `<Iii` struct and it is unsigned int (4bits) + int (4bits) + int (4bits) = 12 bits.

wonjuleee replied:
thanks.

wonjuleee commented on datumaro/plugins/data_formats/datumaro_binary/mapper/media.py:
```diff
@@ -0,0 +1,113 @@
+        elif obj.MEDIA_TYPE == MediaType.MEDIA_ELEMENT:
+            return MediaElementMapper.forward(obj)
+        else:
+            raise DatumaroError(f"{obj.MEDIA_TYPE} is not allowed for MediaMapper.")
```
Could you expect the operation when path is too long to convert string to 8 byte?

vinnamkim replied:
Please see https://github.com/openvinotoolkit/datumaro/pull/830#discussion_r1128804561

vinnamkim commented on datumaro/plugins/data_formats/datumaro_binary/mapper/media.py:
```diff
@@ -0,0 +1,113 @@
+
+        bytes_arr = bytearray()
+        bytes_arr.extend(struct.pack(f"<Iii", cls.MEDIA_TYPE, size[0], size[1]))
+        bytes_arr.extend(StringMapper.forward(obj.path))
```
`path` will be serialized by this so that it can have an arbitrary size.

vinnamkim committed:
Rename save_media() to context_save_media()

Signed-off-by: Kim, Vinnam <vinnam.kim@intel.com>

vinnamkim committed:
Add backward_dict() to MediaElementMapper

 - And ImageMapper and PointCloudMapper inherit MediaElementMapper to
use the common function backward_dict()

Signed-off-by: Kim, Vinnam <vinnam.kim@intel.com>

bonhunko commented on datumaro/plugins/data_formats/datumaro/base.py:
```diff
@@ -136,9 +136,7 @@ def _load_items(self, parsed):
-                            path=osp.join(
-                                self._related_images_dir, self._subset, item_id, ri.get("path")
-                            ),
+                            path=osp.join(ri.get("path")),
```
unnecessary osp.join() call

vinnamkim replied:
Removed.

bonhunko reviewed (dismissed):
Great work!
I left one minor comment.

vinnamkim committed:
Remove unnecessary osp.join()

Signed-off-by: Kim, Vinnam <vinnam.kim@intel.com>
